hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/DecodingFacade.java

```diff
@@ -363,6 +424,179 @@ public Dissociation decodeMultipleDissociations(final Bytes input, final UnaryOp
 		return Dissociation.multiDissociation(accountID, tokenIDs);
 	}
 
+	public TokenCreateWrapper decodeFungibleCreate(final Bytes input, final UnaryOperator<byte[]> aliasResolver) {
+		final Tuple decodedArguments = decodeFunctionCall(input, TOKEN_CREATE_FUNGIBLE_SELECTOR,
+				TOKEN_CREATE_FUNGIBLE_DECODER);
+
+		return decodeTokenCreateWithoutFees(decodedArguments.get(0), true, decodedArguments.get(1),
+				decodedArguments.get(2), aliasResolver);
+	}
+
+	public TokenCreateWrapper decodeFungibleCreateWithFees(
+			final Bytes input,
+			final UnaryOperator<byte[]> aliasResolver
+	) {
+		final Tuple decodedArguments = decodeFunctionCall(input, TOKEN_CREATE_FUNGIBLE_WITH_FEES_SELECTOR,
+				TOKEN_CREATE_FUNGIBLE_WITH_FEES_DECODER);
+
+		final var tokenCreateWrapper = decodeTokenCreateWithoutFees(decodedArguments.get(0), true,
+				decodedArguments.get(1), decodedArguments.get(2), aliasResolver);
+		final var fixedFees = decodeFixedFees(decodedArguments.get(3), aliasResolver);
+		final var fractionalFees = decodeFractionalFees(decodedArguments.get(4), aliasResolver);
+		tokenCreateWrapper.setFixedFees(fixedFees);
+		tokenCreateWrapper.setFractionalFees(fractionalFees);
+
+		return tokenCreateWrapper;
+	}
+
+	public TokenCreateWrapper decodeNonFungibleCreate(final Bytes input, final UnaryOperator<byte[]> aliasResolver
+	) {
+		final Tuple decodedArguments = decodeFunctionCall(input, TOKEN_CREATE_NON_FUNGIBLE_SELECTOR,
+				TOKEN_CREATE_NON_FUNGIBLE_DECODER);
+
+		return decodeTokenCreateWithoutFees(decodedArguments.get(0), false, BigInteger.ZERO, BigInteger.ZERO,
+				aliasResolver);
+	}
+
+	public TokenCreateWrapper decodeNonFungibleCreateWithFees(
+			final Bytes input,
+			final UnaryOperator<byte[]> aliasResolver
+	) {
+		final Tuple decodedArguments = decodeFunctionCall(input, TOKEN_CREATE_NON_FUNGIBLE_WITH_FEES_SELECTOR,
+				TOKEN_CREATE_NON_FUNGIBLE_WITH_FEES_DECODER);
+
+		final var tokenCreateWrapper = decodeTokenCreateWithoutFees(decodedArguments.get(0), false,
+				BigInteger.ZERO, BigInteger.ZERO, aliasResolver);
+		final var fixedFees = decodeFixedFees(decodedArguments.get(1), aliasResolver);
+		final var royaltyFees = decodeRoyaltyFees(decodedArguments.get(2), aliasResolver);
+		tokenCreateWrapper.setFixedFees(fixedFees);
+		tokenCreateWrapper.setRoyaltyFees(royaltyFees);
+
+		return tokenCreateWrapper;
+	}
+
+	private TokenCreateWrapper decodeTokenCreateWithoutFees(
+			final Tuple tokenCreateStruct,
+			final boolean isFungible,
+			final BigInteger initSupply,
+			final BigInteger decimals,
+			final UnaryOperator<byte[]> aliasResolver
+	) {
+		final var tokenName = (String) tokenCreateStruct.get(0);
+		final var tokenSymbol = (String) tokenCreateStruct.get(1);
+		final var tokenTreasury = convertLeftPaddedAddressToAccountId((byte[]) tokenCreateStruct.get(2),
```

Comment:
This works great for building out a `MerkleToken` with `AccountID` treasury and fee collectors.

Not needed for this PR imo, but we probably need to support `ContractID` treasuries and fee collectors (especially because with a contract id `0.0.<create2-address>`, this _can_ actually refer to different `AccountID`s over time if the contract `selfdestruct`s and is then re-created by the same factory).

❔And this raises a second question: Since we prevent token treasuries from being deleted via HAPI, do we also need to prevent a contract treasury from using `selfdestruct`?